samples/matter/common/src/binding_handler.cpp

```diff
@@ -50,6 +50,150 @@ void BindingHandler::OnInvokeCommandFailure(BindingData &aBindingData, CHIP_ERRO
 	}
 }
 
+void BindingHandler::DeviceChangedHandler(const EmberBindingTableEntry &binding, OperationalDeviceProxy *deviceProxy,
+					  void *context)
+{
+	VerifyOrReturn(context != nullptr, LOG_ERR("Invalid context for the device handler"););
+	BindingData *data = static_cast<BindingData *>(context);
+
+	if (binding.type == EMBER_MULTICAST_BINDING && data->IsGroup) {
+		switch (data->ClusterId) {
+#if defined(CONFIG_CHIP_DEVICE_PRODUCT_ID) && (CONFIG_CHIP_DEVICE_PRODUCT_ID == 32772) // 32772 == 0x8004 (example
+										       // light-switch-app)
+		case Clusters::OnOff::Id:
+			OnOffProcessCommand(data->CommandId, binding, nullptr, context);
+			break;
+		case Clusters::LevelControl::Id:
+			LevelControlProcessCommand(data->CommandId, binding, nullptr, context);
+			break;
+#endif // CONFIG_CHIP_DEVICE_PRODUCT_ID
+
+#if defined(CONFIG_THERMOSTAT_EXTERNAL_SENSOR) && (CONFIG_THERMOSTAT_EXTERNAL_SENSOR == true)
+		case Clusters::TemperatureMeasurement::Id:
+			TemperatureMeasurementRead(binding, nullptr, context);
+			break;
+#endif // CONFIG_CHIP_DEVICE_PRODUCT_ID
+
+		default:
+			ChipLogError(NotSpecified, "Invalid binding group command data");
+			break;
+		}
+	} else if (binding.type == EMBER_UNICAST_BINDING && !data->IsGroup) {
+		switch (data->ClusterId) {
+#if defined(CONFIG_CHIP_DEVICE_PRODUCT_ID) && (CONFIG_CHIP_DEVICE_PRODUCT_ID == 32772) // 32772 == 0x8004 (example
+										       // light-switch-app)
+		case Clusters::OnOff::Id:
+			OnOffProcessCommand(data->CommandId, binding, deviceProxy, context);
+			break;
+		case Clusters::LevelControl::Id:
+			LevelControlProcessCommand(data->CommandId, binding, deviceProxy, context);
+			break;
+#endif // CONFIG_CHIP_DEVICE_PRODUCT_ID
+
+#if defined(CONFIG_THERMOSTAT_EXTERNAL_SENSOR) && (CONFIG_THERMOSTAT_EXTERNAL_SENSOR == true)
+		case Clusters::TemperatureMeasurement::Id:
+			TemperatureMeasurementRead(binding, deviceProxy, context);
+			break;
+#endif // CONFIG_CHIP_DEVICE_PRODUCT_ID
+
+		default:
+			ChipLogError(NotSpecified, "Invalid binding unicast command data");
+			break;
+		}
+	}
+}
```

Comment:
This is not a generic code. The BindingHandler API must be refactored to be generic and not contain any specific device types.

Reply:
This is sort of true. Moving a lot of common code will be beneficial for samples anyway. However we could keep the code related to the specific sample in this sample. For example, you could modify Init() method, so it would take callback as an argument, save it in the BindingHandler class as field, implement part of `DeviceChangedHandler` in the samples's `app_task.cpp` and call saved callback once it will be required.